### PR TITLE
Added support for elevation based dark theme overlay color in the Material widget

### PR DESCRIPTION
**Pull request by darrenaustin**
## Description

In order to support the [Material Dark theme specification](https://material.io/design/color/dark-theme.html#properties), this PR adds support for lightening the background color of a `Material` widget when it is in a dark theme based on its elevation.

mio-design_assets_1A1Uf3a9TsbtgSKT-hjIAaEiEIXFDTUfE_darktheme-elevationsystem

If the widget is placed in a dark theme (i.e. `Theme.of(context).brightness == Brightness.dark`) a semi-transparent white will be composited over the `Material.color` for the background of the widget. The level of the transparency follows the values in the specification.

If this feature is not wanted, it can be turned off by the new `darkThemeOverlay` boolean property. By default this is turned on, which may lead to some visual changes to existing apps.

## Tests

I added a group of three new tests to test these features.

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (Please read [Handling breaking changes]). *Link to announcement email TBD*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/material/material.dart [hunk 1]
```diff
@@ -217,9 +221,29 @@ class Material extends StatefulWidget {
...
+  /// If [false] the [color] will be used unmodified.
+  ///
+  /// Defaults to [true].
+  ///
+  /// See also:
+  ///   * [Material.color]
+  ///   * [Material.elevation]
+  ///   * [ThemeData.brightness]
+  ///   * <https://material.io/design/color/dark-theme.html>
+  final bool darkThemeOverlay;
```
Comment (rami-a) [inline packages/flutter/lib/src/material/material.dart]: I think a verb in the bool name is preferred, something like `applyDarkThemeOverlay`

File: packages/flutter/lib/src/material/material.dart [hunk 2]
```diff
@@ -305,16 +330,42 @@ class _MaterialState extends State<Material> with TickerProviderStateMixin {
...
+          elevation <=  1.0 ? 0.05
+        : elevation <=  2.0 ? 0.07
+        : elevation <=  3.0 ? 0.08
+        : elevation <=  4.0 ? 0.09
+        : elevation <=  6.0 ? 0.11
+        : elevation <=  8.0 ? 0.12
+        : elevation <= 12.0 ? 0.14
+        : elevation <= 16.0 ? 0.15
+        : 0.16;
+    final Color whiteOverlay = Color.fromRGBO(255, 255, 255, opacity);
```
Comment (rami-a) [inline packages/flutter/lib/src/material/material.dart]: This could be `Colors.white.withOpacity(opacity)`

**Review by rami-a — APPROVED** [review]
Looks pretty good but this will definitely have visually breaking changes for a lot of people. If anyone is using a dark theme they will have to then opt-out of this.

File: packages/flutter/lib/src/material/material.dart (hunk 1 above)
Comment (darrenaustin) [reply packages/flutter/lib/src/material/material.dart]: Good point. I was also thinking maybe it should include elevation as well. But then `applyDarkThemeElevationOverlay` is really getting to be a mouthful :).

File: packages/flutter/lib/src/material/material.dart (hunk 2 above)
Comment (darrenaustin) [reply packages/flutter/lib/src/material/material.dart]: Yeah, I thought of that, but didn't want to include 'colors.dart' just for `white`.  Looking at it again, that does look more readable.

**Comment by darrenaustin** [comment]
> Looks pretty good but this will definitely have visually breaking changes for a lot of people. If anyone is using a dark theme they will have to then opt-out of this.

Yeah, this concerns me as well. It does seem like it wouldn't be that drastic a visual change, as it only applies a slight white to elevated widgets in a dark theme. That said, I really hate breaking existing apps.

I have thought about making the default off and adding a property to `ThemeData` that would allow us to control it on at the theme level. That would allow a potential `ThemeData.from(colorScheme)` do the right thing here for new apps, but existing apps would be unchanged.

But that leads to more cruft and bloat in `ThemeData` for what may be a small breakage. 

Thoughts?

**Comment by rami-a** [comment]
Hmm I actually do like the idea of adding this to `ThemeData`, despite how bloated it is becoming. 

Otherwise you would have to opt-out for each and every widget you create individually. So even if it is a big breaking change, switching one flag in your `ThemeData` to go back to what you had is a better experience.

I'm still on the fence though about whether it should be opt-in vs opt-out. But I'm definitely in favor of adding the flag to `ThemeData` to give broader control as well as being able to specify on a per-`Material` level.

**Comment by HansMuller** [comment]
@rami-a I agree, it seems safest and least disruptive to make this an opt-in ThemeData flag.

**Comment by darrenaustin** [comment]
I seemed to have borked this branch with some unfortunate git shenanigans. I have opened a new one that has a clean history: #35560.

**Pull request closed without merging**